packages/shared_preferences/shared_preferences/lib/shared_preferences.dart

```diff
@@ -15,34 +15,47 @@ const MethodChannel _kChannel =
 ///
 /// Data is persisted to disk asynchronously.
 class SharedPreferences {
-  SharedPreferences._(this._preferenceCache);
+  SharedPreferences._(this._preferenceCache, {@required this.filename});
 
   static const String _prefix = 'flutter.';
-  static Completer<SharedPreferences> _completer;
+  static final Map<String, Future<SharedPreferences>> _openedInstances =
+      <String, Future<SharedPreferences>>{};
 
-  /// Loads and parses the [SharedPreferences] for this app from disk.
+  /// Returns an instance of [SharedPreferences]
+  /// with values corresponding to those stored under the file with the specified [filename].
   ///
   /// Because this is reading from disk, it shouldn't be awaited in
   /// performance-sensitive blocks.
-  static Future<SharedPreferences> getInstance() async {
-    if (_completer == null) {
-      _completer = Completer<SharedPreferences>();
-      try {
+  ///
+  /// WARNING: [filename] argument for now only works on Android.
+  /// On iOs, the default name will always be used, even with different value in parameter.
+  ///
+  /// The values in [SharedPreferences] are cached.
+  /// A new instance is actually created only the first time this method is called with the specified [filename].
+  ///
+  /// If a file with the specified [filename] doesn't already exist, it will automatically be created.
+  /// The [filename] cannot be null ; otherwise an [ArgumentError] will be thrown.
+  /// The default value of [filename] is the name of the file used in the previous version of this plugin.
+  ///
+  /// For Android, see https://developer.android.com/training/data-storage/shared-preferences.html for more details on the platform implementation.
+  static Future<SharedPreferences> getInstance(
+      {String filename = "FlutterSharedPreferences"}) async {
```

Comment:
I would leave this API alone and define a new one for Android specifically:

`getAndroidPreferencesForFile(String preferencesFile)`

Otherwise, it is misleading. Your application code will potentially contain multiple shared preferences instances and on iOS, these will clobber each other.

Reply:
I created the method `getInstanceForFile`, which will throw an `AssertionError` if the current platform isn't Android. What do you think?

Reply:
Thanks Axel.

The more I think about it, the more I am worried about this API. What is your primary motivation?
- Do you intend to create multiple files backing different shared preferences? 
- or are you in a migration scenario where you have an existing app that is using a different file name and you want to hook shared preferences plugin to it? (and you have no intention to use multiple files?)

The main problem with #1 is that multiple files is not supported on iOS so your app needs to be structured and behave differently on iOS when it comes to preferences. That's a tall order because this is not just a simple switch. For instance, on iOS, if you wanted to separate your preferences, you would need to namespace them. Let's say you were separating preferences for each user.  You would do this on iOS:

`user_id1:preference1`
`user_id1:preference2`
`user_id2:preference1`
`user_id2:preference2`

but then you would write to separate files on Android? Why? You could do the same on Android too and you would not have to worry about differing behavior on different platforms.

Reply:
I apologize for not spotting this earlier. This PR has been open for a while and we did not give proper attention to it.

Reply:
My intention was to use multiple files in Flutter to separate the data. It seemed to me that it could be easier to manage because then you can use functions like `clear` and `getAll` on an entire file without the need to filter the preferences. 
The reason why I didn't code the iOs-native part is because I don't have the knowledge to do so, but with your comment I now understand that it cannot be done at all. So it's probably better to not change the API indeed.